core/services/offchainreporting/OCRContract.go

```diff
@@ -0,0 +1,232 @@
+	if !connected {
+		return nil, errors.New("Failed to register with logBroadcaster")
+	}
+
+	return sub, nil
+}
+
+func (oc *OCRContract) LatestConfigDetails(ctx context.Context) (changedInBlock uint64, configDigest ocrtypes.ConfigDigest, err error) {
+	opts := bind.CallOpts{Context: ctx, Pending: false}
+	result, err := oc.contractCaller.LatestConfigDetails(&opts)
+	if err != nil {
+		return 0, configDigest, errors.Wrap(err, "error getting LatestConfigDetails")
+	}
+	return uint64(result.BlockNumber), ocrtypes.BytesToConfigDigest(result.ConfigDigest[:]), err
+}
+
+func (oc *OCRContract) ConfigFromLogs(ctx context.Context, changedInBlock uint64) (c ocrtypes.ContractConfig, err error) {
+	q := ethereum.FilterQuery{
+		FromBlock: big.NewInt(int64(changedInBlock)),
+		ToBlock:   big.NewInt(int64(changedInBlock)),
+		Addresses: []gethCommon.Address{oc.contractAddress},
+		Topics: [][]gethCommon.Hash{
+			{OCRContractConfigSet},
+		},
+	}
+
+	logs, err := oc.ethClient.FilterLogs(ctx, q)
+	if err != nil {
+		return c, err
+	}
+	if len(logs) == 0 {
+		return c, errors.Errorf("ConfigFromLogs: OCRContract with address 0x%x has no logs", oc.contractAddress)
+	}
+
+	latest, err := oc.contractFilterer.ParseConfigSet(logs[len(logs)-1])
+	if err != nil {
+		return c, errors.Wrap(err, "ConfigFromLogs failed to ParseConfigSet")
+	}
+	latest.Raw = logs[len(logs)-1]
+	return confighelper.ContractConfigFromConfigSetEvent(*latest), err
+}
+
+func (oc *OCRContract) LatestBlockHeight(ctx context.Context) (blockheight uint64, err error) {
+	h, err := oc.ethClient.HeaderByNumber(ctx, nil)
+	if err != nil {
+		return 0, err
+	}
+	if h == nil {
+		return 0, errors.New("got nil head")
+	}
+
+	return uint64(h.Number), nil
+}
+
+func (oc *OCRContract) LatestTransmissionDetails(ctx context.Context) (configDigest ocrtypes.ConfigDigest, epoch uint32, round uint8, latestAnswer ocrtypes.Observation, latestTimestamp time.Time, err error) {
+	opts := bind.CallOpts{Context: ctx, Pending: false}
+	result, err := oc.contractCaller.LatestTransmissionDetails(&opts)
+	if err != nil {
+		return configDigest, 0, 0, ocrtypes.Observation(nil), time.Time{}, errors.Wrap(err, "error getting LatestTransmissionDetails")
+	}
+	return result.ConfigDigest, result.Epoch, result.Round, ocrtypes.Observation(result.LatestAnswer), time.Unix(int64(result.LatestTimestamp), 0), nil
+}
+
+func (oc *OCRContract) FromAddress() gethCommon.Address {
+	return oc.transmitter.FromAddress()
+}
+
+type OCRContractConfigSubscription struct {
+	logger          logger.Logger
+	contractAddress gethCommon.Address
+	ch              chan ocrtypes.ContractConfig
+	oc              *OCRContract
+	mutex           sync.Mutex
+	chClosed        bool
+}
+
+func (sub *OCRContractConfigSubscription) OnConnect() {}
+func (sub *OCRContractConfigSubscription) OnDisconnect() {
+	sub.mutex.Lock()
+	defer sub.mutex.Unlock()
+
+	if !sub.chClosed {
+		sub.chClosed = true
+		close(sub.ch)
```

Comment:
I might be lacking context on this, but I don't think this is what we want here.  `OnDisconnect` is called not just on shutdown but also any time the node temporarily loses connectivity to the ETH node.  LogBroadcaster handles reconnections gracefully and subscribers don't have to care about them.  If you _do_ want to do it this way, it looks like you'd want to move the `sub` initialization code into `OnConnect`.

Reply:
Good to know! I wasn't aware that LogBroadcaster handles that gracefully, so currently, the code that uses the subscription will throw away the subscription and attempt start a new one if `sub.ch` is closed.

I'd be good with the OnDisconnect logic being removed.